src/module-vsbridge-indexer-core/Index/IndexOperations.php

```diff
@@ -68,33 +86,45 @@ class IndexOperations implements IndexOperationInterface
      * @param BulkRequestFactory $bulkRequestFactory
      * @param IndexSettings $indexSettings
      * @param IndexFactory $indexFactory
+     * @param IndexerLogger $indexerLogger
+     * @param OptimizationSettings $optimizationSettings
      */
     public function __construct(
         ClientResolver $clientResolver,
         BulkResponseFactory $bulkResponseFactory,
         BulkRequestFactory $bulkRequestFactory,
         IndexSettings $indexSettings,
-        IndexFactory $indexFactory
+        IndexFactory $indexFactory,
+        IndexerLogger $indexerLogger,
+        OptimizationSettings $optimizationSettings
     ) {
         $this->clientResolver = $clientResolver;
         $this->indexFactory = $indexFactory;
         $this->indexSettings = $indexSettings;
         $this->bulkResponseFactory = $bulkResponseFactory;
         $this->bulkRequestFactory = $bulkRequestFactory;
+        $this->indexerLogger = $indexerLogger;
+        $this->optimizationSettings = $optimizationSettings;
     }
 
     /**
      * @inheritdoc
      */
     public function executeBulk($storeId, BulkRequestInterface $bulk)
     {
+        $this->checkEsCondition($storeId);
+
         if ($bulk->isEmpty()) {
             throw new \LogicException('Can not execute empty bulk.');
         }
 
+        $this->optimizeEsIndexing($storeId, $bulk);
+
         $bulkParams = ['body' => $bulk->getOperations()];
         $rawBulkResponse = $this->resolveClient($storeId)->bulk($bulkParams);
 
+        $this->cleanAfterOptimizeEsIndexing($storeId, $bulk);
+
         return $this->bulkResponseFactory->create(
```

Comment:
I'm not sure if it is the best place to put these optimize functions here. I am not sure it belongs to bulk operation.

What do you think about pushing this to `GenericIndexerHandler`? 
So we have sth like that:
```
$this->indexOperations->optimizeEsIndexing($storeId, $indexName)
 $response = $this->indexOperations->executeBulk($storeId, $bulkRequest);
$this->indexOperations->cleanAfterOptimizeEsIndexing($storeId, $indexName)
```
What do you think about this? @camilloop @jonathanribas 


If you think we should keep this changes in this class, I would change a little structure for BulkRequest class.
and create it like this in `GenericIndexHandler`
```
$this->indexOperations->createBulk($indexName)
$this->indexOperations->createBulk()->addDocuments(
                        $this->typeName,
                        $docs
                    );
```
than you will be able to get index name directly from bulk request, and we will have 100% sure we are performing bulk operation to only one index.

```
      /**
     * @param int $storeId
     * @param BulkRequestInterface $bulk
     *
     * @return BulkResponseInterface
     */
    public function executeBulk($storeId, BulkRequestInterface $bulk)
   { 
        ...
        $this->optimizeEsIndexing($storeId, $bulk->getIndexName());
       ...
   }

  public function optimizeEsIndexing($storeId, $indexName)
    {
      if ($this->optimizationSettings->changeNumberOfReplicas()) {
                $this->resolveClient($storeId)->changeNumberOfReplicas(
                    $indexName,
                    self::NUMBER_OF_REPLICAS_DURING_INDEXING
                );
            }

            if ($this->optimizationSettings->changeRefreshInterval()) {
                $this->resolveClient($storeId)->changeRefreshInterval(
                    $indexName,
                    self::REFRESH_INTERVAL_DURING_INDEXING
                );
            }
}
```

Reply:
It's better to be 100% sure we are performing bulk operation to only one index.

Reply:
@afirlejczyk I was thinking about pushing this code to `GenericIndexerHandler` when i was developing and I think that it's good plac.e for it. So I adapted this code to your first suggestion.